File: Server/src/common/infrastructure/cryptoAlgorithm/OTP.cpp
#include "OTP.h"

const std::string OTP::keyPath = KEYS_PATH + "OTPkey.key";
std::vector<unsigned char> OTP::key;

OTP::OTP()
{
    if (key.size() == 0)
    {
        key = utils::readFileBytes(keyPath);
    }
}

std::string OTP::encrypt(const std::string &message) const
{
    std::string encrypted = "";
    int pos = 0;

    for (const char c : message)
    {
        encrypted += static_cast<char>(c ^ key.at(pos));
        pos++;
        pos %= key.size();
    }

    return encrypted;
}

std::string OTP::decrypt(const std::string &message) const
{
    std::string decrypted = "";
    int pos = 0;

    for (const char c : message)
    {
        decrypted += static_cast<char>(c ^ key.at(pos));
        pos++;
        pos %= key.size();
    }

    return decrypted;
}

File: Client/Trivia/Source/CryptoAlgorithm/OTP.cs
namespace Trivia.CryptoAlgorithm;

//REVIEW: Consider renaming to Otp
public class OTP : ICryptoAlgorithm
{
    private const string KeyPath = ICryptoAlgorithm.KeysPath + "OTPkey.key";
    
    private static byte[] _key = [];

    public OTP()
    {
        if (_key.Length == 0)
        {
            _key = ICryptoAlgorithm.ReadFileBytes(KeyPath);
        }
    }

    public string Encrypt(string message)
    {
        var encrypted = "";
        var pos = 0;
        
        foreach (var c in message)
        {
            encrypted += (char)(c ^ _key[pos]);
            pos++;
            pos %= _key.Length;
        }
        
        return encrypted;
    }

    public string Decrypt(string message)
    {
        var decrypted = "";
        var pos = 0;
        
        foreach (var c in message)
        {
            decrypted += (char)(c ^ _key[pos]);
            pos++;
            pos %= _key.Length;
        }

        return decrypted;
    }
}
